### PR TITLE
Add google analytics pipeline (ga4)

### PR DESCRIPTION
**Pull request by redicane**

### CODE REVIEW

File: pipelines/google_analytics/helpers/credentials.py [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+        """
+        Will produce an access token from the given credentials.
+        """
+        try:
+            google = OAuth2Session(client_id=self.client_id, scope=["https://www.googleapis.com/auth/analytics.readonly"], redirect_uri="https://localhost")
+            extra = {
+                "client_id": self.client_id,
+                "client_secret": self.client_secret
+            }
+            self.access_token = google.refresh_token(token_url="https://oauth2.googleapis.com/token", refresh_token=self.refresh_token, **extra)["access_token"]
```
Comment (rudolfix) [inline pipelines/google_analytics/helpers/credentials.py:48]: please add scopes and redirect url to config

**Review by rudolfix — CHANGES_REQUESTED** [review]
This is really close. Please rebase on newest master and fix the poetry lock conflict (you can just add the dependency again)

File: pipelines/google_analytics/__init__.py [hunk 2]
```diff
@@ -0,0 +1,159 @@
...
+    :param property_id: A reference to the Google Analytics project. https://developers.google.com/analytics/devguides/reporting/data/v1/property-id
+    :param rows_per_page: Controls how many rows are retrieved per page in the reports. Default is 10000, maximum possible is 100000.
+    :param resource_name: The resource name used to save incremental into dlt state
+    :param start_date: Incremental load start_date, default is taken from dlt state if exists
+    :param end_date: Incremental load end_date, default is always set to the date when the load is started.
+    :returns: Generator of all rows of data in the report.
+    """
+
+    # grab the start time from last dlt load if not filled, if that is also empty then use the first day of the millennium as the start time instead
+    if not start_date:
```
[the remark this answers is not in the document]
Comment (rudolfix) [reply pipelines/google_analytics/__init__.py:103]: why do you need state here? what is the purpose to restart from previous date? if there is no real use for let's remove it. also @adrianbr did you request that?

File: pipelines/google_analytics/__init__.py [hunk 3]
```diff
@@ -0,0 +1,163 @@
...
+    # Build the service object for Google Analytics api.
+    client = BetaAnalyticsDataClient(credentials=credentials)
+    # get metadata needed for some resources
+    metadata = get_metadata(client=client, property_id=property_id)
+    resource_list = [metadata | metrics_table, metadata | dimensions_table]
+    for query in queries:
+        name = query["resource_name"]
+        dimensions = query["dimensions"]
+        metrics = query["metrics"]
+        resource_list.append(dlt.resource(basic_report(client=client, rows_per_page=rows_per_page, property_id=property_id, dimensions=dimensions, metrics=metrics,
```
Comment (rudolfix) [inline pipelines/google_analytics/__init__.py:81]: do we really want to replace reports? if you do them daily I assume you want to keep daily reports. what is a typical use case @adrianbr ? running this daily and keep reports for given day?

**Review by rudolfix — CHANGES_REQUESTED** [review]

File: pipelines/google_analytics/__init__.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (redicane) [reply pipelines/google_analytics/__init__.py:103]: Hey @rudolfix, 

The idea here was to have the chance to run this periodically (like with Zendesk) and not have to input start and end dates manually. It is just extra functionality on loading the resource.

**Comment by adrianbr** [comment]
Common use case we need to support
1. All requests should include date unless they already include DateHour or DateHourMinute
2. We should not load partial dates such as Today (today is not complete)
3. we do not need selected periods loaded. Incremental loading should always load from start date until last complete period (yesterday)
4. IMO we should not start with merge but with append loading. Load from start/default/last val to yesterday.
5. Merge loading is based on rules that may not apply to GA4 - let's do it once we have specs from a user such as Hans. Merge loading would always replace last X weeks of data to refresh data for which events arrived late. I am not sure this is still an issue in GA4 and 2023.

**Comment by rudolfix** [comment]
@adrianbr I think it is not clear how incremental loading should work from pipeline user perspective. we'll have identical problems in all pipelines like that

> we do not need selected periods loaded. Incremental loading should always load from start date until last complete period (yesterday)

what does it mean? to me this means that we replace all the reports in the destination each time pipeline is run. and the start date is explicitly provided. if this is not the case then tell me

right now pipeline appends the data. if start date is specified it generates reports from start data until now and then appends
if it is not specified it generates the report form the beginning (here year 2001 AFAIK) and appends them. on next run it will use the 'yesterday' from the last run. so this is the incremental part. is this what you want? so to lead incrementally you never specify a start data. is that what you want?

i'm not able to review this further. also I see that dates can still be 'today' but we need to clear this first

**Comment by adrianbr** [comment]
@rudolfix yes, this is how we want it loaded.

In short, it should load incrementally from
- If it loaded before: last loaded date
- if it did not load before: 
-   If there is a start date, start from there. Else, start from beginning of time. (year 2000 is fine)

**Comment by rudolfix** [comment]
maybe you can write pseudo code how people would load incrementally with start date specified

**Pull request merged**